File: hphp/hack/src/hhbc/emit_record_def.ml
open Core_kernel
open Aast
open Hhbc_string_utils

let valid_tc_for_record_field tc =
  match Hhas_type_constraint.name tc with
  | None -> true
  | Some name ->
    (not (is_self name))
    && (not (is_parent name))
    && (not (String.lowercase name = "hh\\this"))
    && (not (String.lowercase name = "callable"))
    && (not (String.lowercase name = "hh\\nothing"))
    && not (String.lowercase name = "hh\\noreturn")

let emit_field namespace (f : Aast.sid * Aast.hint * Tast.expr option) =
  let ((pos, name), hint, expr) = f in
  let otv =
    match expr with
    | Some expr -> Ast_constant_folder.expr_to_opt_typed_value namespace expr
    | None -> None
  in
  let tc =
    Emit_type_hint.hint_to_type_info
      ~kind:Emit_type_hint.Property
      ~nullable:false
      ~skipawaitable:false
      ~tparams:[]
      hint
  in
  if valid_tc_for_record_field (Hhas_type_info.type_constraint tc) then
    (name, tc, otv)
  else
    Emit_fatal.raise_fatal_parse
      pos
      (Printf.sprintf
         "Invalid record field type hint for '%s'"
         (Utils.strip_ns name))

let emit_record_def rd =
  let elaborate (_, name) = Hhbc_id.Record.from_ast_name name in
  let name = elaborate rd.rd_name in
  let parent_name =
    match rd.rd_extends with
    | Some (_, Aast.Happly (name, _)) -> Some name
    | _ -> None
  in
  let parent_name = Option.map ~f:elaborate parent_name in
  Hhas_record_def.make
    name
    rd.rd_abstract
    parent_name
    (List.map rd.rd_fields ~f:(emit_field rd.rd_namespace))

let emit_record_defs_from_program
    (ast : (Closure_convert.hoist_kind * Tast.def) list) =
  let aux (_, d) =
    match d with
    | RecordDef rd -> Some (emit_record_def rd)
    | _ -> None
  in
  List.filter_map ~f:aux ast

File: hphp/hack/src/hhbc/emit_record_def.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use ast_constant_folder_rust as constant_folder;
use emit_fatal_rust as emit_fatal;
use emit_type_hint_rust as emit_type_hint;
use env::emitter::Emitter;
use hhas_pos_rust::Span;
use hhas_record_def_rust::{Field as RecordField, HhasRecord};
use hhas_type::constraint;
use hhbc_id_rust::record;
use instruction_sequence::Result;
use oxidized::namespace_env::Env as Namespace;

use hhbc_string_utils_rust as string_utils;
use oxidized::ast::*;

fn valid_tc_for_record_field(tc: &constraint::Type) -> bool {
    match &tc.name {
        None => true,
        Some(name) => {
            !(string_utils::is_self(name)
                || string_utils::is_parent(name)
                || name.eq_ignore_ascii_case("hh\\this")
                || name.eq_ignore_ascii_case("callable")
                || name.eq_ignore_ascii_case("hh\\nothing")
                || name.eq_ignore_ascii_case("hh\\noreturn"))
        }
    }
}

fn emit_field<'a>(
    emitter: &Emitter,
    namespace: &Namespace,
    field: &'a (Sid, Hint, Option<Expr>),
) -> Result<RecordField<'a>> {
    let (Id(pos, name), hint, expr_opt) = field;
    let otv = expr_opt.as_ref().map_or(None, |e| {
        constant_folder::expr_to_typed_value(emitter, namespace, e).ok()
    });
    let ti = emit_type_hint::hint_to_type_info(
        &emit_type_hint::Kind::Property,
        false,
        false,
        &[],
        &hint,
    )?;
    if valid_tc_for_record_field(&ti.type_constraint) {
        Ok(RecordField(name.as_str(), ti, otv))
    } else {
        let name = string_utils::strip_global_ns(name);
        Err(emit_fatal::raise_fatal_parse(
            &pos,
            format!("Invalid record field type hint for '{}'", name),
        ))
    }
}

fn emit_record_def<'a>(emitter: &Emitter, rd: &'a RecordDef) -> Result<HhasRecord<'a>> {
    fn elaborate<'b>(Id(_, name): &'b Id) -> record::Type<'b> {
        name.trim_start_matches("\\").into()
    }
    let parent_name = match &rd.extends {
        Some(Hint(_, h)) => {
            if let Hint_::Happly(id, _) = h.as_ref() {
                Some(elaborate(id))
            } else {
                None
            }
        }
        _ => None,
    };
    Ok(HhasRecord {
        name: elaborate(&rd.name),
        is_abstract: rd.abstract_,
        base: parent_name,
        fields: rd
            .fields
            .iter()
            .map(|f| emit_field(emitter, rd.namespace.as_ref(), &f))
            .collect::<Result<Vec<_>>>()?,
        span: Span::from_pos(&rd.span),
    })
}

pub fn emit_record_defs_from_program<'a>(
    emitter: &Emitter,
    p: &'a Program,
) -> Result<Vec<HhasRecord<'a>>> {
    p.into_iter()
        .filter_map(|d| d.as_record_def().map(|r| emit_record_def(emitter, r)))
        .collect::<Result<Vec<_>>>()
}
